src/Stratis.Bitcoin/Consensus/ProvenHeadersConsensusManagerBehavior.cs

```diff
@@ -0,0 +1,253 @@
+        private async Task ProcessGetProvenHeadersAsync(INetworkPeer peer, GetProvenHeadersPayload getProvenHeadersPayload)
+        {
+            if (getProvenHeadersPayload.BlockLocator.Blocks.Count > BlockLocator.MaxLocatorSize)
+            {
+                this.logger.LogTrace("Peer '{0}' sent getprovenheader with oversized locator, disconnecting.", peer.RemoteSocketEndpoint);
+
+                peer.Disconnect("Peer sent getprovenheaders with oversized locator");
+
+                this.logger.LogTrace("(-)[LOCATOR_TOO_LARGE]");
+                return;
+            }
+
+            // Ignoring "getprovenheaders" from peers because node is in initial block download unless the peer is whitelisted.
+            // We don't want to reveal our position in IBD which can be used by attacker. Also we don't won't to deliver peers any blocks
+            // because that will slow down our own syncing process.
+            if (this.initialBlockDownloadState.IsInitialBlockDownload() && !peer.Behavior<IConnectionManagerBehavior>().Whitelisted)
+            {
+                this.logger.LogTrace("(-)[IGNORE_ON_IBD]");
+                return;
+            }
+
+            ProvenHeadersPayload headersPayload = this.ConstructProvenHeadersPayload(getProvenHeadersPayload.BlockLocator, getProvenHeadersPayload.HashStop, out ProvenBlockHeader lastHeader);
+
+            if (headersPayload != null)
+            {
+                this.logger.LogTrace("{0} headers were selected for sending, last one is '{1}'.", headersPayload.Headers.Count, headersPayload.Headers.LastOrDefault()?.GetHash());
+
+                try
+                {
+                    this.BestSentHeader = lastHeader;
+
+                    await peer.SendMessageAsync(headersPayload).ConfigureAwait(false);
+                }
+                catch (OperationCanceledException)
+                {
+                    this.logger.LogTrace("Unable to send headers message to peer '{0}'.", peer.RemoteSocketEndpoint);
+                }
+            }
+        }
+
+        /// <summary>
+        /// Processes "headers" message received from the peer.
+        /// </summary>
+        /// <param name="peer">Peer from which the message was received.</param>
+        /// <param name="provenHeadersPayload">Payload of "Proven Headers" message to process.</param>
+        /// <remarks>
+        /// "provenHeaders" message is sent in response to "getprovenheaders" message or it is solicited
+        /// by the peer when a new block is validated (unless in IBD).
+        /// <para>
+        /// When we receive "provenheaders" message from the peer, we can adjust our knowledge
+        /// of the peer's view of the chain. We update its pending tip, which represents
+        /// the tip of the best chain we think the peer has.
+        /// </para>
+        /// </remarks>
+        protected async Task ProcessProvenHeadersAsync(INetworkPeer peer, ProvenHeadersPayload provenHeadersPayload)
```

Comment:
no need to make it protected. 
also the method copypasted from `ProcessHeadersAsync`, please get rid of copy-pasted code